rpc/src/runtime.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::methods::{build_registry, RpcRegistry};
+use futures::future::join_all;
+use libra_config::config::NodeConfig;
+use serde::Serialize;
+use serde_json::map::Map;
+use serde_json::Value;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use storage_client::{StorageRead, StorageReadServiceClient};
+use tokio::runtime::{Builder, Runtime};
+use warp::Filter;
+
+/// Creates HTTP server (warp-based) that serves JSON RPC requests
+/// Returns handle to corresponding Tokio runtime
+pub fn bootstrap(address: SocketAddr, storage_client: Arc<dyn StorageRead>) -> Runtime {
+    let runtime = Builder::new()
+        .thread_name("rpc-")
+        .threaded_scheduler()
+        .enable_all()
+        .build()
+        .expect("[rpc] failed to create runtime");
+
+    let registry = Arc::new(build_registry());
+
+    let handler = warp::any()
+        .and(warp::path::end())
+        .and(warp::post())
+        .and(warp::header::exact("content-type", "application/json"))
+        .and(warp::body::json())
+        .and(warp::any().map(move || Arc::clone(&storage_client)))
+        .and(warp::any().map(move || Arc::clone(&registry)))
+        .and_then(rpc_endpoint);
+
+    let future = warp::serve(handler).run(address);
+    runtime.handle().spawn(future);
+    runtime
+}
+
+/// Creates JSON RPC endpoint by given node config
+pub fn bootstrap_from_config(config: &NodeConfig) -> Runtime {
+    let storage_client: Arc<dyn StorageRead> =
+        Arc::new(StorageReadServiceClient::new(&config.storage.address));
+
+    bootstrap(config.rpc.address, storage_client)
+}
+
+/// JSON RPC entry point
+/// Handles all incoming rpc requests
+/// Performs routing based on methods defined in `registry`
+async fn rpc_endpoint(
+    data: Value,
+    storage_client: Arc<dyn StorageRead>,
+    registry: Arc<RpcRegistry>,
+) -> Result<Box<dyn warp::Reply>, warp::Rejection> {
+    if let Value::Array(requests) = data {
+        // batch API call
+        let futures = requests.into_iter().map(|req| {
+            rpc_request_handler(req, Arc::clone(&storage_client), Arc::clone(&registry))
+        });
+        let responses = join_all(futures).await;
+        Ok(Box::new(warp::reply::json(&Value::Array(responses))))
+    } else {
+        // single API call
+        let resp = rpc_request_handler(data, storage_client, registry).await;
+        Ok(Box::new(warp::reply::json(&resp)))
+    }
+}
+
+/// Handler of single RPC request
+/// Performs validation and executes corresponding rpc handler
+async fn rpc_request_handler(
+    req: Value,
+    storage_client: Arc<dyn StorageRead>,
+    registry: Arc<RpcRegistry>,
+) -> Value {
+    let request: Map<String, Value>;
+    let mut response = Map::new();
+
+    // set defaults: protocol version to 2.0, request id to null
```

Comment:
what does this mean? the comment could be made more useful to state why we're doing this or making a readme that describes how our json-rpc interface works.